keps/sig-cloud-provider/20191004-out-of-tree-credential-providers.md

```diff
@@ -0,0 +1,207 @@
+type RegistryCredentialProvider struct {
+    metav1.TypeMeta `json:",inline"`
+
+    // ImageURLMatch is used to match against the Image URL to determine which images to
+    // provide credentials for.
+    //
+    // The ImageURLMatch string is a URL pattern which can optionally contain a port and a path.  Globs can be used in the hostname (but not the port or the path).
+    //
+    // An image URL matches when it has the same number of parts as the ImageURLMatch, and each part matches.  Additionally the path of ImageURLMatch must be a prefix of the target URL. If the ImageURLMatch contains a port, then the port must match as well.
+    ImageURLMatch []string
+
+    // Exec specifies a custom exec-based plugin.  This type is defined in
+    // https://github.com/kubernetes/client-go/blob/62f256057db7571c5ed1aba47eea291f72dd557a/tools/clientcmd/api/types.go#L184
+    Exec clientcmd.ExecConfig
+}
+```
+
+The RegistryCredentialProvider will be encoded in JSON and located in a file on disk.  The directory where credential provider configuration is found will be passed to kubelet via a new configuration option `RegistryCredentialProviderDir`.
+
+We will create new types `RegistryCredentialPluginSpec` and `RegistryCredentialPluginStatus` which will define the interface between the plugin and the kubelet runtime.  After the kubelet matches the image URL to a RegistryCredentialProvider, the kubelet will exec the plugin binary, and pass the JSON encoded spec to the plugin via stdin.  This includes the image URL that is to be pulled.  The plugin will report back the status, which includes the credentials that kubelet needs to pull the image.
+
+In the in-tree implementation, the docker keyring, which has N credential providers, returns an `[]AuthConfig` on a Lookup(image string) call.  This struct will be populated by the plugin rather than the in-tree provider.
+
+```go
+// RegistryCredentialPluginSpec is passed to the plugin via stdin, and includes the image that will be pulled by kubelet.
+type RegistryCredentialPluginSpec struct {
+    // Image is used when passed to registry credential providers as part of an image pull
+    // +optional
+    Image string `json:"image,omitempty"`
+}
+
+// RegistryCredentialPluginStatus holds credentials for the kubelet runtime to use for image pulls.  It is returned from the plugin as stdout.
+type RegistryCredentialPluginStatus struct {
+    // +optional
+    ExpirationTimestamp *metav1.Time `json:"expirationTimestamp,omitempty"`
+
+    // +optional
+    Username *string `json:"username,omitempty"`
+    // +optional
+    Password *string `json:"password,omitempty"`
+
+    // IdentityToken is used to authenticate the user and get
+    // an access token for the registry.
+    // +optional
+    IdentityToken *string `json:"identitytoken,omitempty"`
+
+    // RegistryToken is a bearer token to be sent to a registry
+    // +optional
+    RegistryToken *string `json:"registrytoken,omitempty"`
+}
+
+```
+
```

Comment:
I think it would be super helpful to show real examples (in yaml?) of `RegistryCredentialProvider`, `RegistryCredentialPluginSpec` and `RegistryCredentialPluginStatus` that we would use for one of the existing providers (ECR?)